### PR TITLE
Optimize `ecrecover` ASM

### PR DESCRIPTION
**Pull request by wborgeaud**
Optimize ECDSA verification assembly by using GLV + wNAF + MSM. 
An average ETH (valid) tx signature takes 37678 cycles to verify (currently 90089).

### CODE REVIEW

File: evm/src/cpu/kernel/asm/curve/secp256k1/ecrecover.asm [hunk 1]
```diff
@@ -62,70 +62,58 @@ ecrecover_valid_input:
...
+    %stack (bit_a1, bit_b0, bit_b1, i, accx, accy, a0, a1, b0, b1, retdest) -> (i, a0, bit_a1, bit_b0, bit_b1, i, accx, accy, a0, a1, b0, b1, retdest)
+    SHR %and_const(1)
+    %mul_const(2) ADD %mul_const(2) ADD %mul_const(2) ADD
+    %stack (index, i, accx, accy, a0, a1, b0, b1, retdest) -> (index, index, i, accx, accy, a0, a1, b0, b1, retdest)
+    %mul_const(2) %add_const(1)
+    %mload_kernel(@SEGMENT_KERNEL_ECDSA_TABLE)
+    SWAP1 %mul_const(2)
+    %mload_kernel(@SEGMENT_KERNEL_ECDSA_TABLE)
+    %stack (Px, Py, i, accx, accy, a0, a1, b0, b1, retdest) -> (Px, Py, accx, accy, ecdsa_after_precompute_loop_contd, i, a0, a1, b0, b1, retdest)
+    %jump(ec_add_valid_points_secp)
```
Comment (unzvfu) [inline evm/src/cpu/kernel/asm/curve/secp256k1/ecrecover.asm:106]: I think I said this elsewhere, but you know that the two input points are non-zero here, so the checks at the start of `ec_add_valid_points_secp` are unnecessary.

**Review by unzvfu — APPROVED** [review]
Nice work! Good to go, modulo a few minor things.

Some notes on things that aren't part of the pull request:

- nit: You can remove most of the `SWAP`s in `ecrecover` by changing the order of `DUP`s and `PUSH`s; something like `%ecrecover_input_check; PUSH 27; SWAP3; SUB; DUP4; %secp_liftx` should suffice.
- `ecrecover_input_check` verifies that both `r` and `s` are less than `secp256k1n`. This is what appears in, for example, [some documents](https://www.evm.codes/precompiled#0x01?fork=merge), however the [Yellow paper](https://ethereum.github.io/yellowpaper/paper.pdf) says that `s` should be less than `secp256k1n / 2 + 1`. I suggest we work out which one it is and document the choice in the code.
- Line 60 of `ecrecover.asm`, looks like it should be `SWAP6` not `SWAP5`?!
- Comment on line 18 of `curve_add.asm` should say 'second' instead of 'first' (but note review comments about how those checks might not be necessary)

File: evm/src/cpu/kernel/asm/curve/secp256k1/curve_add.asm [hunk 2]
```diff
@@ -198,42 +198,32 @@ ec_add_equal_points:
     %jump(ec_add_valid_points_with_lambda)
 
 // Secp256k1 elliptic curve doubling.
-// Assumption: (x0,y0) is a valid point.
+// Assumption: (x,y) is a valid point.
 // Standard doubling formula.
 global ec_double_secp:
```
[the remark this answers is not in the document]
Comment (wborgeaud) [reply evm/src/cpu/kernel/asm/curve/secp256k1/curve_add.asm:203]: Good point, I'll address this in another PR.

File: evm/src/cpu/kernel/asm/curve/secp256k1/curve_add.asm [hunk 3]
```diff
@@ -198,42 +198,32 @@ ec_add_equal_points:
...
-    // stack: x0, y0, retdest
-    DUP2
-    // stack: y0, x0, y0, retdest
-    DUP2
-    // stack: x0, y0, x0, y0, retdest
+    // stack: x, y, retdest
+    DUP2 DUP2 %ec_isidentity
+    // stack: (x,y)==(0,0), x, y, retdest
```
[the remark this answers is not in the document]
Comment (wborgeaud) [reply evm/src/cpu/kernel/asm/curve/secp256k1/curve_add.asm:206]: Hm I'm not sure we can assume the input isn't the identity. It's pretty easy to craft adversarial inputs that make the first few iterations of the MSM loop double the zero point.

Same thing for `ec_add_valid_points_secp`, in which case we also add the zero point in the loop 1/16 times on average.

File: evm/src/cpu/kernel/asm/curve/secp256k1/curve_add.asm [hunk 4]
```diff
@@ -198,42 +198,32 @@ ec_add_equal_points:
...
-    // stack: x < x - y, x - y
-    %secp_base
-    // stack: N, x < x - y, x - y
-    MUL
-    // stack: N * (x < x - y), x - y
-    ADD
-    // (x-y) % N
+    %stack (x, y) -> (@SECP_BASE, y, x, @SECP_BASE)
+    SUB ADDMOD
 %endmacro
```
[the remark this answers is not in the document]
Comment (wborgeaud) [reply evm/src/cpu/kernel/asm/curve/secp256k1/curve_add.asm:227]: Is it? It's not in `OpsColumnsView` or `TernaryOperator` at least.

File: evm/src/cpu/kernel/asm/curve/secp256k1/precomputation.asm [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    // stack: Qx, @SECP_BASE, Qx, Qy, retdest
+    PUSH @SECP_GLV_BETA MULMOD
+    %stack (betaQx, Qneg, Qx, Qy, retdest) -> (Qneg, Qy, Qneg, betaQx, Qx, Qy, retdest)
+    MUL SWAP1 PUSH 1 SUB
+    // stack: 1-Qneg, Qneg*Qy, betaQx, Qx, Qy, retdest
+    DUP5 PUSH @SECP_BASE SUB MUL ADD
+    // SUB MUL SWAP1 DUP5 PUSH @SECP_BASE SUB MUL ADD
+    %stack (selectQy, betaQx, Qx, Qy, retdest) -> (2, betaQx, 3, selectQy, betaQx, selectQy, Qx, Qy, precompute_table_contd, retdest)
+    %mstore_kernel(@SEGMENT_KERNEL_ECDSA_TABLE) %mstore_kernel(@SEGMENT_KERNEL_ECDSA_TABLE)
+    %jump(ec_add_valid_points_secp)
```
[the remark this answers is not in the document]
Comment (wborgeaud) [reply evm/src/cpu/kernel/asm/curve/secp256k1/precomputation.asm:30]: Good point, I changed it for this addition. For the additions below, I think we might hit some edge cases with some adversarial inputs.

File: evm/src/cpu/kernel/asm/curve/secp256k1/precomputation.asm [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    DUP1 %increment %mload_kernel(@SEGMENT_KERNEL_ECDSA_TABLE)
+    %stack (y, i, retdest) -> (i, y, i, retdest)
+    %mload_kernel(@SEGMENT_KERNEL_ECDSA_TABLE)
+    PUSH precompute_table_loop_contd
+    DUP3 DUP3
+    PUSH 9 %mload_kernel(@SEGMENT_KERNEL_ECDSA_TABLE)
+    PUSH 8 %mload_kernel(@SEGMENT_KERNEL_ECDSA_TABLE)
+    // stack: Gx, Gy, x, y, precompute_table_loop_contd, x, y, i, retdest
+    %jump(ec_add_valid_points_secp)
+global precompute_table_loop_contd:
```
[the remark this answers is not in the document]
Comment (wborgeaud) [reply evm/src/cpu/kernel/asm/curve/secp256k1/precomputation.asm:47]: Oops, just used the global for debugging but they should just be local.

File: evm/src/cpu/kernel/asm/curve/secp256k1/ecrecover.asm (hunk 1 above)
Comment (wborgeaud) [reply evm/src/cpu/kernel/asm/curve/secp256k1/ecrecover.asm:106]: `Px,Py` can be `0,0` here.

**Review by wborgeaud — COMMENTED** [review]
> * nit: You can remove most of the `SWAP`s in `ecrecover` by changing the order of `DUP`s and `PUSH`s; something like `%ecrecover_input_check; PUSH 27; SWAP3; SUB; DUP4; %secp_liftx` should suffice.

Good catch. This code predates the introduction of `%stack`, so I've just changed it to use it.

> * `ecrecover_input_check` verifies that both `r` and `s` are less than `secp256k1n`. This is what appears in, for example, [some documents](https://www.evm.codes/precompiled#0x01?fork=merge), however the [Yellow paper](https://ethereum.github.io/yellowpaper/paper.pdf) says that `s` should be less than `secp256k1n / 2 + 1`. I suggest we work out which one it is and document the choice in the code.

So the `s < secp256k1n / 2 + 1` is for signature verification on the network. For ecrecover, the condition is indeed `s < secp256k1n` (see equation (210) in the Yellow Paper). Don't ask me why they are different...

> * Line 60 of `ecrecover.asm`, looks like it should be `SWAP6` not `SWAP5`?!

Hm why? `SWAPi` swaps the 1st and `i+1`th element of the stack. Here we want to swap the 1st and 6th (`s`) element of the stack.
```
// stack: r^(-1), r^(-1), y, hash, x, s, retdest
SWAP5
// stack: s, r^(-1), y, hash, x, r^(-1), retdest
```

> * Comment on line 18 of `curve_add.asm` should say 'second' instead of 'first' (but note review comments about how those checks might not be necessary)

👍

File: evm/src/cpu/kernel/asm/curve/secp256k1/curve_add.asm (hunk 4 above)
[the remark this answers is not in the document]
Comment (unzvfu) [reply evm/src/cpu/kernel/asm/curve/secp256k1/curve_add.asm:220]: Oh right, the constraints have been included but it's not tied up to the rest of the system. You'll have to wait a bit longer...

**Comment by unzvfu** [comment]
> > * Line 60 of `ecrecover.asm`, looks like it should be `SWAP6` not `SWAP5`?!
> 
> Hm why? `SWAPi` swaps the 1st and `i+1`th element of the stack. Here we want to swap the 1st and 6th (`s`) element of the stack.

Oops, you're right, ignore that comment.

File: evm/src/cpu/kernel/asm/curve/secp256k1/ecrecover.asm (hunk 1 above)
Comment (unzvfu) [reply evm/src/cpu/kernel/asm/curve/secp256k1/ecrecover.asm:94]: Ah, yes, I see. I was thinking that all table entries should be non-zero, but of course an adversary could select `P = -G` etc.

File: evm/src/cpu/kernel/asm/curve/secp256k1/curve_add.asm (hunk 3 above)
[the remark this answers is not in the document]
Comment (unzvfu) [reply evm/src/cpu/kernel/asm/curve/secp256k1/curve_add.asm:192]: Yep you're right, I hadn't thought through the adversarial inputs fully.

**Pull request merged**